### PR TITLE
Ga 47 create table access types

### PR DESCRIPTION
**Pull request by patelila**
### Jira link (if applicable)

https://tools.hmcts.net/jira/browse/GA-47

### Change description ###

PRM : Create table access_types

### CODE REVIEW

**Comment by helenbirdhmcts** [comment]
Is there a test for "monotonically increasing version number which is incremented each time a new version of the data is written" or are we manually testing this is working correctly?

**Comment by patelila** [comment]
> Is there a test for "monotonically increasing version number which is incremented each time a new version of the data is written" or are we manually testing this is working correctly?

**Comment by patelila** [comment]
I tested manually that it is increasing everytime an insert is done as shown in ticket.

File: src/main/resources/db/migration/V20231218_47__GA-47_new_access_types_table.sql [hunk 1]
```diff
@@ -0,0 +1,14 @@
+-- create table
+-- PK/FK relationships in tables access_types
+-- access_types pk column = version
+-- automatic when creating bigserial:
+--   access_types_version_seq (Sequence)
+--   nextval('access_types_version_seq') NOT null
+
+CREATE TABLE access_types(
+	version bigserial not null,
+    access_type jsonb NOT NULL,
```
Comment (abul-syed) [inline src/main/resources/db/migration/V20231218_47__GA-47_new_access_types_table.sql:10]:
```suggestion
    access_types jsonb NOT NULL,
```

should be `access_types` - jira is incorrect, should also be updated

File: src/main/resources/db/migration/V20231218_47__GA-47_new_access_types_table.sql [hunk 2]
```diff
@@ -0,0 +1,14 @@
+-- create table
```
Comment (abul-syed) [inline src/main/resources/db/migration/V20231218_47__GA-47_new_access_types_table.sql:1]: comments left over?

File: src/main/java/uk/gov/hmcts/reform/orgrolemapping/data/AccessTypes.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    @Serial
+    private static final long serialVersionUID = 1L;
+
+    @Id
+    @Column(name = "version")
+    @GeneratedValue(strategy = GenerationType.IDENTITY, generator = "version_seq")
+    private Long version;
+
+    @Column(name = "access_type", nullable = false)
+    private String accessType;
```
Comment (abul-syed) [inline src/main/java/uk/gov/hmcts/reform/orgrolemapping/data/AccessTypes.java:36]:
```suggestion
    @Column(name = "access_types", nullable = false)
    private String accessTypes;
```

see docs: https://tools.hmcts.net/confluence/pages/viewpage.action?pageId=1723368280#PRMMVP-3.7.5.2CaseDefinitionAccessTypes

File: src/main/java/uk/gov/hmcts/reform/orgrolemapping/data/AccessTypesRepository.java [hunk 4]
```diff
@@ -0,0 +1,12 @@
+package uk.gov.hmcts.reform.orgrolemapping.data;
+
+import org.springframework.data.repository.CrudRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface AccessTypesRepository extends CrudRepository<AccessTypes, Long> {
+
+    AccessTypes findByAccessType(String json);
```
Comment (abul-syed) [inline src/main/java/uk/gov/hmcts/reform/orgrolemapping/data/AccessTypesRepository.java:9]: not sure if `findByAccessType` will be used. but can always remove later

**Review by abul-syed — CHANGES_REQUESTED** [review]
few comments

File: src/main/resources/db/migration/V20231218_47__GA-47_new_access_types_table.sql (hunk 1 above)
Comment (sanjayparekh-hmcts) [reply src/main/resources/db/migration/V20231218_47__GA-47_new_access_types_table.sql:10]: agree with Abul - need to clear up with Suman

File: src/main/resources/db/migration/V20231218_47__GA-47_new_access_types_table.sql (hunk 2 above)
Comment (patelila) [reply src/main/resources/db/migration/V20231218_47__GA-47_new_access_types_table.sql:1]: Comments are there to say that PK/FK relationship is automatically created.

**Review by helenbirdhmcts — CHANGES_REQUESTED** [review]

File: src/main/resources/db/migration/V20231218_47__GA-47_new_access_types_table.sql [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+-- PK/FK relationships in tables access_types
+-- access_types pk column = version
+-- automatic when creating bigserial:
+--   access_types_version_seq (Sequence)
+--   nextval('access_types_version_seq') NOT null
+
```
[the remark this answers is not in the document]
Comment (helenbirdhmcts) [reply src/main/resources/db/migration/V20231218_47__GA-47_new_access_types_table.sql:7]: agreed these are useful comments, so leave these

**Review by helenbirdhmcts — APPROVED** [review]

**Review by sanjayparekh-hmcts — DISMISSED** [review]

File: src/main/java/uk/gov/hmcts/reform/orgrolemapping/data/AccessTypesRepository.java [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+@Repository
+public interface AccessTypesRepository extends CrudRepository<AccessTypes, Long> {
+
+    //retrieves the AccessTypes record by version.
+    AccessTypes findByVersion(Long version);
+
+    // Retrieves the first AccessTypes record.
+    // 1. It first sorts the AccessTypes records in descending order by version.
+    // 2.Hence, we get the top version among all the AccessTypes
+    AccessTypes findFirstByOrderByVersionDesc();
```
Comment (abul-syed) [inline src/main/java/uk/gov/hmcts/reform/orgrolemapping/data/AccessTypesRepository.java:15]: `findFirstByOrderByVersionDesc()` is sorting required, if there will only be 1 row? see: https://tools.hmcts.net/confluence/pages/viewpage.action?pageId=1723368280#PRMMVP-3.7.5.2CaseDefinitionAccessTypes `findFirst()` should be fine?

Comment (patelila) [reply src/main/java/uk/gov/hmcts/reform/orgrolemapping/data/AccessTypesRepository.java:15]: There are lots of rows in the table versioned. So to return only the last version, yes it is necessary to sort so the latest is the first row and that is what gets returned. See the ticket as I have added a manual test of this.

**Review by sanjayparekh-hmcts — DISMISSED** [review]

**Review by helenbirdhmcts — DISMISSED** [review]

**Review by abul-syed — APPROVED** [review]

**Review by abul-syed — APPROVED** [review]

**Review by sanjayparekh-hmcts — APPROVED** [review]

**Review by mattnayler — COMMENTED** [review]
GA-47/DTSAM-296 Reviewed alongside #1737.

**Pull request merged**